core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/workflow/WorkflowRewriter.scala

```diff
@@ -0,0 +1,424 @@
+package edu.uci.ics.texera.workflow.common.workflow
+
+import com.typesafe.scalalogging.Logger
+import edu.uci.ics.texera.workflow.common.storage.OpResultStorage
+import edu.uci.ics.texera.Utils.objectMapper
+import edu.uci.ics.texera.workflow.common.operators.OperatorDescriptor
+import edu.uci.ics.texera.workflow.common.workflow.WorkflowRewriter.copyOperator
+import edu.uci.ics.texera.workflow.operators.sink.CacheSinkOpDesc
+import edu.uci.ics.texera.workflow.operators.source.cache.CacheSourceOpDesc
+
+import java.util.UUID
+import scala.collection.mutable
+
+case class WorkflowVertex(
+    op: OperatorDescriptor,
+    links: mutable.HashSet[OperatorLink]
+)
+
+object WorkflowRewriter {
+  private def copyOperator(operator: OperatorDescriptor): OperatorDescriptor = {
+    objectMapper.readValue(objectMapper.writeValueAsString(operator), classOf[OperatorDescriptor])
+  }
+}
+
+class WorkflowRewriter(
+    var workflowInfo: WorkflowInfo,
+    var cachedOperatorDescriptors: mutable.HashMap[String, OperatorDescriptor],
+    var cacheSourceOperatorDescriptors: mutable.HashMap[String, CacheSourceOpDesc],
+    var cacheSinkOperatorDescriptors: mutable.HashMap[String, CacheSinkOpDesc],
+    var operatorRecord: mutable.HashMap[String, WorkflowVertex],
+    var opResultStorage: OpResultStorage
+) {
+
+  private val logger = Logger(this.getClass.getName)
+
+  var visitedOpID: mutable.HashSet[String] = new mutable.HashSet[String]()
+
+  private val workflowDAG: WorkflowDAG = if (workflowInfo != null) {
+    new WorkflowDAG(workflowInfo)
+  } else {
+    null
+  }
+  private val rewrittenToCacheOperatorIDs = if (null != workflowInfo) {
+    new mutable.HashSet[String]()
+  } else {
+    null
+  }
+  private val r1NewOps = if (workflowInfo != null) {
+    mutable.MutableList[OperatorDescriptor]()
+  } else {
+    null
+  }
+  private val r1NewLinks = if (workflowInfo != null) {
+    mutable.MutableList[OperatorLink]()
+  } else {
+    null
+  }
+  private val r1NewBreakpoints = if (workflowInfo != null) {
+    mutable.MutableList[BreakpointInfo]()
+  } else {
+    null
+  }
+  private val r2NewOps = if (workflowInfo != null) {
+    mutable.MutableList[OperatorDescriptor]()
+  } else {
+    null
+  }
+  private val r2NewLinks = if (workflowInfo != null) {
+    mutable.MutableList[OperatorLink]()
+  } else {
+    null
+  }
+  private val r2NewBreakpoints = if (workflowInfo != null) {
+    mutable.MutableList[BreakpointInfo]()
+  } else {
+    null
+  }
+  private val r2OpIDQue = if (workflowInfo != null) {
+    new mutable.Queue[String]()
+  } else {
+    null
+  }
+
+  var r1WorkflowInfo: WorkflowInfo = _
+  var r1WorkflowDAG: WorkflowDAG = _
+  var r2WorkflowInfo: WorkflowInfo = _
+  var r2WorkflowDAG: WorkflowDAG = _
```

Comment:
what do `r1` and `r2` stand for? please rename it to make it more readable, or provide some comments?